### PR TITLE
Delay deletion of pod from the API server until volumes are deleted

### PR DESCRIPTION
**Pull request by dashpole**
Depends on #37228, and will not pass tests until that PR is merged, and this is rebased.

Keeps all kubelet behavior the same, except the kubelet will not make the "Delete" call (kubeClient.Core().Pods(pod.Namespace).Delete(pod.Name, deleteOptions)) until the volumes associated with that pod are removed.  I will perform some performance testing so that we better understand the latency impact of this change.

Is kubelet_pods.go the correct file to include the "when can I delete this pod" logic?

cc: @vishh @sjenning @derekwaynecarr

### CODE REVIEW

**Comment by vishh** [comment]
> Is kubelet_pods.go the correct file to include the "when can I delete this pod" logic?

It can be in `kubernetes/pkg/kubelet` go package for now. File name is up to you.

**Comment by vishh** [comment]
This PR mostly LGTM. Can you add the e2e test you created for this purpose to this PR?

**Comment by dashpole** [comment]
@vishh  This change is tested by any test that creates or deletes pods.  I created the test to make testing my change quicker, as it runs in ~1 minute.  I can add it if you would like, but it is largely redundant.

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -675,6 +676,38 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return false
 }
 
+func (kl *Kubelet) safeToDeletePod(pod *v1.Pod) bool {
```
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: i thought @vishh wanted to pend on cgroup deletion as well?

File: pkg/kubelet/status/status_manager.go [hunk 2]
```diff
@@ -66,6 +66,8 @@ type manager struct {
 	// Map from (mirror) pod UID to latest status version successfully sent to the API server.
 	// apiStatusVersions must only be accessed from the sync thread.
 	apiStatusVersions map[types.UID]uint64
+	// A function which returns true if the pod can saftely be deleted
```
Comment (derekwaynecarr) [inline pkg/kubelet/status/status_manager.go]: nit: safely

**Review by derekwaynecarr — COMMENTED** [review]
thanks for doing this, just the one question on pending until cgroup is cleaned up as well, we could just block on PodContainerManager.Exists(pod)

**Comment by derekwaynecarr** [comment]
i would prefer to put this function in `kubelet_pods.go`  (cc @pmorie ) .

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: add some godoc here explaining what and why this exists as it will not be obvious to future folks... i know for a fact the relationship around the whole delete pod flow in kubelet is black magic to many.

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -675,6 +676,38 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
...
+		// We shouldnt delete pods whose DeletionTimestamp is not set
+		return false
+	}
+	if !notRunning(pod.Status.ContainerStatuses) {
+		// We shouldnt delete pods that still have running containers
+		glog.V(3).Infof("Pod %q is terminated, but some containers are still running", format.Pod(pod))
+		return false
+	}
+	if kl.podVolumesExist(pod.UID) && !kl.kubeletConfiguration.KeepTerminatedPodVolumes {
+		// We shouldnt delete pods whose volumes have not been cleaned up if we are not keeping terminated pod volumes
```
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: is it worth logging here or is there sufficient logging elsewhere if/when the volume could not be cleaned up (suspect there is, but worth confirming)

**Comment by gnufied** [comment]
cc @kubernetes/sig-storage-misc

**Comment by vishh** [comment]
We might want to trigger deletion of volumes by posting a signal to the
volume manager using channels for example. Not necessary for this PR, but
we should watch out for test timeouts once this PR goes in.

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (dashpole) [inline pkg/kubelet/kubelet_pods.go]: Yes, we should also do that.  Is: "kl.containerManager.NewPodContainerManager().Exists(pod)" the correct statement?

File: pkg/kubelet/status/status_manager.go (hunk 2 above)
Comment (dashpole) [inline pkg/kubelet/status/status_manager.go]: done.

**Comment by sjenning** [comment]
@dashpole https://github.com/kubernetes/kubernetes/pull/37228 merged about 15 minutes after you posted this :)  You can rebase now to see if the tests pass.  The PR lgtm.

**Comment by yujuhong** [comment]
> Keeps all kubelet behavior the same, except the kubelet will not make the "Delete" call (kubeClient.Core().Pods(pod.Namespace).Delete(pod.Name, deleteOptions)) until the volumes associated with that pod are removed.

When/how will kubelet trigger the pod deletion after the volumes are cleaned up?

File: pkg/kubelet/kubelet_pods.go (hunk 3 above)
Comment (dashpole) [inline pkg/kubelet/kubelet_pods.go]: There is logging related to volume unmounting in [pkg/kubelet/volumemanager/reconciler/reconciler.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go).  My only concern is that most debugging people wont make the connection that their pod is not being deleted because unmounting a volume failed.  But I'm not sure how to notify users without logging every time they try to delete a pod (which is useless).

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (dashpole) [inline pkg/kubelet/kubelet_pods.go]: Done.

**Comment by vishh** [comment]
Sorry, I failed to post my comments earlier.

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: that looks right to me.

File: pkg/kubelet/kubelet_pods.go [hunk 4]
```diff
@@ -675,6 +676,38 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return false
 }
 
+func (kl *Kubelet) safeToDeletePod(pod *v1.Pod) bool {
+	if kubepod.IsMirrorPod(pod) {
```
Comment (dashpole) [inline pkg/kubelet/kubelet_pods.go]: I dont have full context on how mirror pods work.  This is taken directly from the old status manager code.

File: pkg/kubelet/status/status_manager.go [hunk 5]
```diff
@@ -66,6 +66,8 @@ type manager struct {
...
+	safeToDeletePod func(pod *v1.Pod) bool
```
Comment (dashpole) [inline pkg/kubelet/status/status_manager.go]: Great idea.  I will change that.

File: pkg/kubelet/kubelet_pods.go [hunk 6]
```diff
@@ -675,6 +676,38 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
...
+	if pod.DeletionTimestamp == nil {
+		// We shouldnt delete pods whose DeletionTimestamp is not set
+		return false
+	}
+	if !notRunning(pod.Status.ContainerStatuses) {
+		// We shouldnt delete pods that still have running containers
+		glog.V(3).Infof("Pod %q is terminated, but some containers are still running", format.Pod(pod))
+		return false
+	}
+	if kl.podVolumesExist(pod.UID) && !kl.kubeletConfiguration.KeepTerminatedPodVolumes {
```
Comment (dashpole) [inline pkg/kubelet/kubelet_pods.go]: added.

**Comment by yujuhong** [comment]
> When/how will kubelet trigger the pod deletion after the volumes are cleaned up?

I spoke to @dashpole offline. We need to verify that the deletion will still be triggered eventually before we can merge this PR.

**Comment by dashpole** [comment]
> When/how will kubelet trigger the pod deletion after the volumes are cleaned up?

This is an issue, and I am exploring a couple solutions, and making sure I understand the problem.

There actually is a test just like the one I made: test/e2e_node/simple_mount.go

I havent been able to get cleaning up cgroups.  

Based on my tests, cleaning up volumes adds ~2 seconds to pod deletion time.  Is this acceptable?

**Comment by dashpole** [comment]
All node e2e tests passed when I ran it locally, ill have to take a look at why that e2e test failed.

**Comment by dashpole** [comment]
@k8s-bot gci gce e2e test this

**Comment by dashpole** [comment]
@k8s-bot node e2e test this

**Review by vishh — APPROVED** [review]

**Comment by dashpole** [comment]
Note that this does not include cgroup cleanup.  I tried adding it in, but it was not working.  I am not sure if that is because it increases latency even more, or if it has other issues.  I can handle that in a separate PR.

**Comment by vishh** [comment]
That's ok!

**Comment by dashpole** [comment]
Another note is that this now checks to see if an existing pod can be deleted as part of syncBatch in the status manager.  This may add up to 10 seconds of latency in deletion, because syncBatch is run on a 10 second timer.

**Comment by dashpole** [comment]
@k8s-bot bazel test this

**Comment by dashpole** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot bazel test this

**Comment by dashpole** [comment]
See #40824 for my first round of profiling (excluding this change).  It appears that volume cleanup occurs almost immediately after deletion.  Ill do some profiling including this change to see how the graph changes.

**Comment by dashpole** [comment]
This change appears to add about 60 seconds of additional latency for pod deletion.
Here is the new profile:
screenshot from 2017-02-02 10 57 53

This is the comparison of the 50th, 90th, and 99th percentile of latency.  All three increase almost exactly 60 seconds from the previous value.
screenshot from 2017-02-02 11 00 04

This is surprisingly high, and I plan to look into why that is.

**Comment by jingxu97** [comment]
@dashpole, just want to check whether you are still working on this PR? I reviewed a PR recently which might depends on this one. Thanks!

**Comment by dashpole** [comment]
I am still working on this.  I think I have found why it added so much delay, but I need to confirm that.

**Comment by dashpole** [comment]
Thanks to @Random-Liu for helping me debug this.
This graph shows deletion INCLUDING this change with a QPS limit of 5.
screenshot from 2017-02-06 11 00 58

This graph shows deletion INCLUDING this change with a QPS limit of 60.
screenshot from 2017-02-06 11 01 53

As I have discussed with @vishh in the past, this change currently requires an additional round trip:

1.  Deletion timestamp set via API call to kubelet; pod manager updated to include deletion timestamp (DARK GREY)
2.  Containers killed, causes status update to the API server (RED)
3.  ROUND TRIP to API server
4.  Containers terminated reflected back to kubelet via API call; pod manager updated to include terminated containers
5.  Volume Manager tears down volumes, as it sees that containers are terminated via pod manager. (LIGHT GREY)
6.  Pod is Deleted from the API server. (PURPLE)

The problem that I did not realize existed is that round trips to the API server take an enormous amount of time when trying to update/delete 105 pods due to QPS constraints.  This is easy to see when comparing the first and second graphs.

**Comment by dashpole** [comment]
To verify my theory, I did the following:

- In Step 2 above, I also updated the pod manager to reflect that the containers had died

The following are the same as the graphs above, but including this change.

QPS = 5
screenshot from 2017-02-06 13 14 44

QPS = 60
screenshot from 2017-02-06 13 14 27

It is clear that this change eliminates entirely the gap between containers being killed (RED) and volumes being cleaned up (LIGHT GREY).

**Comment by dashpole** [comment]
Since messing with the syncloop is a bad idea, I plan to modify the volume manager to use the status manager instead of the pod manager to determine when it can clean up a pod.  This should have the same effect as the syncloop change since it eliminates the roundtrip to the API server

**Comment by jingxu97** [comment]
@dashpole Could you explain more about your fix?

Since messing with the syncloop is a bad idea, I plan to modify the volume manager to use the status manager instead of the pod manager to determine when it can clean up a pod. This should have the same effect as the syncloop change since it eliminates the roundtrip to the API server

Which "syncloop" you mean in the above comment? What is the status manager?

**Comment by dashpole** [comment]
The syncloop is the core kubelet thread that handles all updates to pod state.  I wanted to avoid modifying it because it can have a lot of unforeseen impacts.  The status manager is a kubelet component that is responsible for processing pod status updates, and propagates those updates to the API server.  The pod manager is another kubelet component that keeps internal state about pods.  It is updated when the kubelet receives a pod from the API server.  In general, it is better to refer to the status manager for the most up-to-date pod status.  However, the volume manager was checking to see if a pod was terminated based on the pod manager.  This meant that in order for the volume manager to clean up a pod, the status change indicating containers had died had to propegate: -> status manager -> api server -> pod manager -> volume manager.
My change is to make the volume manager check the status manager when checking to see if a pod is terminated: ->status manager -> volume manager.  This eliminates the round trip from the previous behavior.

**Comment by dashpole** [comment]
I am investigating a bug possibly produced by this PR.  I see the following in the kubelet logs during runs that are >2 minutes long:
`Feb 13 20:22:02 test-gci-dev-57-9202-20-0 kubelet[7670]: I0213 20:22:02.242720    7670 status_manager.go:439] Status for pod "pod-with-prestop-exec-hook_e2e-tests-container-lifecycle-hook-3b0pj(07a0b6d4-f22a-11e6-9fae-42010a800002)" updated successfully: {status:{Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614102 nsec:0 loc:0x49cad80}} Reason: Message:} {Type:Ready Status:False LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614122 nsec:0 loc:0x49cad80}} Reason:ContainersNotReady Message:containers with unready status: [pod-with-prestop-exec-hook]} {Type:PodScheduled Status:True LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614103 nsec:0 loc:0x49cad80}} Reason: Message:}] Message: Reason: HostIP:10.128.0.2 PodIP: StartTime:0xc420fd7f00 InitContainerStatuses:[] ContainerStatuses:[{Name:pod-with-prestop-exec-hook State:{Waiting:<nil> Running:<nil> Terminated:0xc420ed2700} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:false RestartCount:0 Image:gcr.io/google_containers/busybox:1.24 ImageID:docker-pullable://gcr.io/google_containers/busybox@sha256:4bdd623e848417d96127e16037743f0cd8b528c026e9175e22a84f639eca58ff ContainerID:docker://cacdcea4166da97894fb0ba3fd49ee02aef4dd4fd90f3d050983ab21556df30a}] QOSClass:BestEffort} version:5 podName:pod-with-prestop-exec-hook podNamespace:e2e-tests-container-lifecycle-hook-3b0pj}`

`Feb 13 20:22:02 test-gci-dev-57-9202-20-0 kubelet[7670]: I0213 20:22:02.243002    7670 kubelet_pods.go:745] Pod "pod-with-prestop-exec-hook_e2e-tests-container-lifecycle-hook-3b0pj(07a0b6d4-f22a-11e6-9fae-42010a800002)" is terminated, but some volumes have not been cleaned up`

`Feb 13 20:22:02 test-gci-dev-57-9202-20-0 kubelet[7670]: I0213 20:22:02.250713    7670 status_manager.go:439] Status for pod "pod-with-prestop-exec-hook_e2e-tests-container-lifecycle-hook-3b0pj(07a0b6d4-f22a-11e6-9fae-42010a800002)" updated successfully: {status:{Phase:Running Conditions:[{Type:Initialized Status:True LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614102 nsec:0 loc:0x49cad80}} Reason: Message:} {Type:Ready Status:True LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614122 nsec:0 loc:0x49cad80}} Reason: Message:} {Type:PodScheduled Status:True LastProbeTime:{Time:{sec:0 nsec:0 loc:0x49cad80}} LastTransitionTime:{Time:{sec:63622614103 nsec:0 loc:0x49cad80}} Reason: Message:}] Message: Reason: HostIP:10.128.0.2 PodIP:10.180.0.246 StartTime:0xc420fd7f00 InitContainerStatuses:[] ContainerStatuses:[{Name:pod-with-prestop-exec-hook State:{Waiting:<nil> Running:0xc420bd38e0 Terminated:<nil>} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:true RestartCount:0 Image:gcr.io/google_containers/busybox:1.24 ImageID:docker-pullable://gcr.io/google_containers/busybox@sha256:4bdd623e848417d96127e16037743f0cd8b528c026e9175e22a84f639eca58ff ContainerID:docker://cacdcea4166da97894fb0ba3fd49ee02aef4dd4fd90f3d050983ab21556df30a}] QOSClass:BestEffort} version:6 podName:pod-with-prestop-exec-hook podNamespace:e2e-tests-container-lifecycle-hook-3b0pj}
Feb 13 20:22:02 test-gci-dev-57-9202-20-0 kubelet[7670]: I0213 20:22:02.250793    7670 kubelet_pods.go:740] Pod "pod-with-prestop-exec-hook_e2e-tests-container-lifecycle-hook-3b0pj(07a0b6d4-f22a-11e6-9fae-42010a800002)" is terminated, but some containers are still running` 

These lines are followed by a long period of inaction on the pod, and finally a syncloop that updates the status to failed (correct), volumes are unmounted, and then pod deletion.

To summarize:
`Set Pod Status to Failed`
`Cant delete pod because volumes are still mounted`
`Set Pod Status to Running`
`..... long period of time ~1-2 minutes .......`
`syncloop sync pod`
`Set Pod Status to Failed`
`Unmount Pod's volumes`
`Successfully delete Pod`

**Comment by dashpole** [comment]
I have been unable to replicate this on head, so it must be a problem introduced by this PR

**Comment by dashpole** [comment]
Mystery solved!
